drivers: bluetooth: hci: add support for bt_disable() function call

nirav-agrawal opened the pull request:
- Added flag in nxp setup driver to perform HCI controller firmware- load and configuration only once. This avoids loading/setting-up controller firmware on successive h4_open() after h4_close().
- Added HCI command to send HCI_RESET in `bt_disable()` function and remove its call from hci-driver close() call.
- Remove rsp buf pass to `hci_reset_complete()` function as status is already checked under 'bt_hci_cmd_send_sync()' func.

For HCI H4 driver, close() call is added in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/86424.
one more PR to be merged before current PR, https://github.com/zephyrproject-rtos/zephyr/pull/86462/files

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +548,32 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+
+	LOG_DBG("");
+
+	/* Reset the Controller */
+	if (IS_ENABLED(CONFIG_BT_HCI_HOST)) {
+		ret = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, NULL);
```
This doesn't look like something that the HCI driver should be responsible for, rather the host should do it straight in `bt_disable()` (likely conditional to the no-reset quirk). I can see this has slipped into some other drivers' `close()` implementations, like `ipc.c` and `hci_nxp.c` so as part of adding the reset to the host side it should be removed from those places.

nirav-agrawal replied:
Hi @jhedberg, thank you for your quick review. Sure, I will move this code in HCI Core under bt_disable () flow. I just checked the other driver implementation and tried to align similar to it. But you are correct, it should be taken care by the stack function. Thank you.

nirav-agrawal replied:
Hi @jhedberg, changes are made as suggested. pls check. thank you.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4382,6 +4383,18 @@ int bt_disable(void)
 
+	/* Reset the Controller */
+	if (!drv_quirk_no_reset()) {
+		struct net_buf *rsp;
+		err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, &rsp);
+		if (err) {
+			LOG_ERR("Failed to reset BLE controller");
+			return err;
+		}
+		hci_reset_complete(rsp);
+		net_buf_unref(rsp);
```
```suggestion
		struct net_buf *rsp;
		
		err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, &rsp);
		if (err) {
			LOG_ERR("Failed to reset BLE controller");
			return err;
		}
		
		hci_reset_complete(rsp);
		net_buf_unref(rsp);
```

nirav-agrawal replied:
Hi @cvinayak , suggestion amended. thank you.

cvinayak reviewed (changes requested):
Does this duplicate or supercede https://github.com/zephyrproject-rtos/zephyr/pull/86424 ?

axelnxp commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -507,19 +507,14 @@ static int bt_nxp_close(const struct device *dev)
 	int ret = 0;
-	/* Reset the Controller */
-	if (IS_ENABLED(CONFIG_BT_HCI_HOST)) {
-		ret = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, NULL);
```
I'm not ok with this change, you are impacting all our NXP platforms including RW61x, MCXW7x
I'm not sure why you change this file, your PR only mentions H4 interface, this is not an H4 interface.

nirav-agrawal replied:
Hi @axelnxp, This PR is related to add close() function in hci driver to support bt_disable() where for some driver, the close() support was not present. 

As part of this change, HCI Reset was happening from driver, but not from the bt_disable() function.  As per the review comment given by @jhedberg, HCI Reset command is removed from the driver during close() call because now the stack would take care on sending same command before closing HCI instance. 

Please describe if any implications it would have without hci-reset command call inside driver?
 
Regards,
Nirav

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4382,6 +4383,20 @@ int bt_disable(void)
 
+	/* Reset the Controller */
+	if (!drv_quirk_no_reset()) {
+		struct net_buf *rsp;
+
+		err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, &rsp);
```
I just realized that we have some dead code in `hci_reset_complete()`. The `bt_hci_cmd_send_sync()` will return an error if it got a non-zero HCI status, and the `rsp` pointer will not be set then. So both places now calling this should just remove the buffer completely.

nirav-agrawal replied:
Hi @jhedberg, thanks for your review. I split commit into two. first commit represents hci_reset related change, and rsp buffer changes. second commit represent driver change to support close() to open() calls and avoid unnecessory firmware-load and setups.

Let me know if you are expecting any more change. 

For H4 driver, I reverted my change which adds h4_close() function because there is already a PR with similar changes. https://github.com/zephyrproject-rtos/zephyr/pull/86424. 

Thank you.

jhedberg replied:
Your commit message still talks about `h4_close()`: "drivers: bluetooth: hci: control fw-load/setup to support h4_close()". The remaining NULL check also doesn't make much sense without the actual `close()` implementation, so IMO these should be moved into the same PR (maybe put everything into this one?).

Minor thing about commit message subject prefixes: `drivers: bluetooth hci:` does provide as much context as it could. I'd propose to use `drivers: bluetooth: h4: ...` to make it clear what driver you're targeting. Regarding the one for `hci_core.c` I think that's the main context there and the driver changes are only side-effects of the host-side change, so I'd use something like "Bluetooth: Host: Perform HCI_Reset in bt_disable()" for its subject.

nirav-agrawal replied:
Hi @jhedberg, for commit msg related changes, I updated PR. pls confirm. 

For H4 driver related changes, initially I had added close() function support in current PR, but then it is discovered that the similar PR is already raised where h4_close() is added. https://github.com/zephyrproject-rtos/zephyr/pull/86424/commits

PR#86424 is raised by [lfilliot](https://github.com/lfilliot), hence I removed relevant changes from my PR and only kept required once. 

Thank you,

nirav-agrawal replied:
Ok, I will keep track of #86424, once it gets merged, current PR will go in. 
Thank you so much @jhedberg for your time and review.

nirav-agrawal committed:
bluetooth: host: perform hci-reset in bt_disable()

- added HCI command to send hci-reset in bt_disable() func, and
  remove its call from the driver close() call.
- remove rsp buf pass to 'hci_reset_complete()' func as status is
  already checked under 'bt_hci_cmd_send_sync()'.

Signed-off-by: Nirav Agrawal <nirav.agrawal@nxp.com>

nirav-agrawal committed:
drivers: bluetooth: hci_nxp_setup: support for bt disable and reinit

- added flag in nxp setup driver to perform HCI controller firmware-
  load and configuration only once. This avoid loading/setting-up
  controller fw on successive bt_init() after bt_disable() call.

Signed-off-by: Nirav Agrawal <nirav.agrawal@nxp.com>

zhaynxp commented on drivers/bluetooth/hci/hci_nxp.c:
```diff
@@ -507,19 +507,14 @@ static int bt_nxp_close(const struct device *dev)
+
+	ret = PLATFORM_TerminateBle();
+	if (ret < 0) {
+		LOG_ERR("Failed to shutdown BLE controller");
```
Hi @nirav-agrawal,

This change conflicts with our change in https://github.com/zephyrproject-rtos/zephyr/pull/86462/files, is it ok that only perform HCI_RESET, not power off CPU2 when bt disable? Since we're seeing some CPU1 access issue in SMU region in coex scenario.

nirav-agrawal replied:
Hi @zhaynxp, the only change I did was to remove hci- reset calls. In this code, the above code was called as per return status of hci-reset command, so I updated code to remove that dependency. 
I understand your point; I will wait for your PR to be merged and will rebase my PR to keep your change. It will be look like below code, 

`static int bt_nxp_close(const struct device *dev)
{
	struct bt_nxp_data *hci = dev->data;

	hci->recv = NULL;

	return 0;
}`

please suggest. thank you. 

regards,
Nirav

axelnxp reviewed (approved):
ok with the change, just remember this will conflict with #86462 and will need sequencing

nirav-agrawal commented:
HI @axelnxp, @jhedberg, 
I rebased this PR with latest main where dependent PR is already merged. Pls approve this PR to get merged. Thank you.

jhedberg commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -256,8 +257,10 @@ static void rx_thread(void *p1, void *p2, void *p3)
 			uart_irq_rx_enable(cfg->uart);
 
-			LOG_DBG("Calling bt_recv(%p)", buf);
-			h4->recv(dev, buf);
+			if (h4->recv != NULL) {
+				LOG_DBG("Calling bt_recv(%p)", buf);
+				h4->recv(dev, buf);
+			}
```
`h4_close()` calls `k_thread_abort(cfg->rx_thread);` before setting `h4->recv = NULL;`. I'm not sure how thread aborting works in practice, but I'd assume that the right thing to do here in case `recv` is NULL is to just return from the function, right?

nirav-agrawal replied:
Hi @jhedberg, Yes your point is valid. if we abort the thread then no need to put such condition. I will revert those change and update this PR. thank you

nirav-agrawal replied:
Hi @jhedberg, the changes in h4.c is reverted. please check. thank you.

nirav-agrawal commented:
Hi @jhedberg, @dleach02, can you please approve this PR if nothing pending for review?
thank you.

nirav-agrawal commented:
Hi @cvinayak, can you please review again and approve this PR. thank you
